### ISSUE TITLE
PubSub out of order confirmations

### ISSUE THREAD

**Issue by rueian** [dragonflydb/dragonfly#1245]
Hi, congratulations on the release of 1.3.0! I am very excited about the new release.

However, I found there is an out-of-order issue for the SUBSCRIBE/PSUBSCRIBE responses while testing the latest release with my RESP3 client (https://github.com/redis/rueidis) by the following code:

```go
package main

import (
	"context"
	"fmt"
	"github.com/redis/rueidis"
)

func main() {
	client, err := rueidis.NewClient(rueidis.ClientOption{
		InitAddress:  []string{"127.0.0.1:6379"},
		DisableCache: true,
	})
	if err != nil {
		panic(err)
	}
	defer client.Close()
	for _, resp := range client.DoMulti(
		context.Background(),
		client.B().Set().Key("key1").Value("1").Build(),
		client.B().Get().Key("key1").Build(),
		client.B().Subscribe().Channel("ch1").Build(),
		client.B().Set().Key("key2").Value("2").Build(),
		client.B().Get().Key("key2").Build(),
		client.B().Psubscribe().Pattern("ch2").Build(),
	) {
		fmt.Println(resp.ToString())
	}
}

```

This code works will pipeline the following commands to the server in RESP3:
```
SET key1 1
GET key1
SUBSCRIBE ch1
SET key2 2
GET key2
PSUBSCRIBE ch2
```

and it works well with the official redis but easily get failed on the dragonfly.

That is because the dragonfly sends responses in an out-of-order way:

```
redis:
+ OK
$ 1
> subscribe ch1
+ OK
$ 2
> psubscribe ch2
```
```
dragonfly:
+ OK
$ 1
+ OK
$ 2
> psubscribe ch2
> subscribe ch1
```

I believe the out-of-order behavior is a bug because a SUBSCRIBE/PSUBSCRIBE/SSUBSCRIBE command may still result in an ERR response and the correct order is the only way to check if the command succeeded.

**Comment by romange** [comment]
@dranikpg  probably due to tx squashing. Can you please take a look?

@rueian  I see you use `DoMulti` for performance reasons, but you could just use pipelining (without multi/exec). 
Do you also need atomicity semantics that come with `MULTI` ?

**Comment by rueian** [comment]
Hey @romange,

`DoMulti` indeed just uses pipelining without multi/exec.

I think in terms of the above case, there is no need for multi/exec. It is just a common scenario that normal commands and pubsub commands are mixed together in a RESP3 connection.

However, since you mentioned, I also tried with multi/exec:

```go
package main

import (
	"context"
	"fmt"
	"github.com/redis/rueidis"
)

func main() {
	client, err := rueidis.NewClient(rueidis.ClientOption{
		InitAddress:  []string{"127.0.0.1:6379"},
		DisableCache: true,
	})
	if err != nil {
		panic(err)
	}
	defer client.Close()
	tx, err := client.DoMulti(
		context.Background(),
		client.B().Multi().Build(),
		client.B().Set().Key("key1").Value("1").Build(),
		client.B().Get().Key("key1").Build(),
		client.B().Subscribe().Channel("ch1").Build(),
		client.B().Set().Key("key2").Value("2").Build(),
		client.B().Get().Key("key2").Build(),
		client.B().Psubscribe().Pattern("ch2").Build(),
		client.B().Exec().Build(),
	)[7].ToArray()
	if err != nil {
		panic(err)
	}
	for _, resp := range tx {
		fmt.Println(resp)
	}
}
```

The above code sends:

```
MULTI
SET key1 1
GET key1
SUBSCRIBE ch1
SET key2 2
GET key2
PSUBSCRIBE ch2
EXEC
```
and here is what a redis responds with:
```
redis:
+ OK
+ QUEUED
+ QUEUED
+ QUEUED
+ QUEUED
+ QUEUED
+ QUEUED
*
  + OK
  $ 1
  > subscribe ch1
  + OK
  $ 2
  > psubscribe ch2
```
and here is what a dragondly responds with:
```
redis:
+ OK
+ QUEUED
+ QUEUED
+ QUEUED
+ QUEUED
+ QUEUED
+ QUEUED
*
  + OK
  $ 1
  + OK
  $ 2
  > psubscribe ch2
  > subscribe ch1
```

**Comment by romange** [comment]
Thank you @rueian ! We will look into it.

**Comment by dranikpg** [comment]
I'm working on a fix, but I need to check what the actual behavior of subscribe replies is in multi/exec context. Its really confusing 😕 

@rueian 
It seems that rueidis isn't sure as well 🙂 If subscribe to two channels at once, then it crashes (protocol bug panic) with Redis.

```go
for _, resp := range client.DoMulti(
		context.Background(),
		client.B().Multi().Build(),
		client.B().Set().Key("key1").Value("1").Build(),
		client.B().Get().Key("key1").Build(),
		client.B().Subscribe().Channel("ch1", "ch2").Build(),
		client.B().Set().Key("key2").Value("2").Build(),
		client.B().Get().Key("key2").Build(),
		client.B().Psubscribe().Pattern("ch3").Build(),
		client.B().Exec().Build(),
	) {
		fmt.Println(resp.ToString())
	}
```

@romange 

This is how subscribe works at a glance:
> When successful, this command doesn't return anything. Instead, for each channel, one message with the first element being the string "subscribe" is pushed as a confirmation that the command succeeded.

Now if I send this to Redis:
```
MULTI
SUBSCRIBE foo1 foo2 foo3
ECHO first
ECHO second
EXEC
```

I get three acks from subscribe, but the echos are ignored. Why? Because EXEC expects every command to produce **exactly one output** item - but subscribe itself generates 3 and thus emplaces the other replies.

On the other hand, `SUBSCRIBE` can't delay its output, because again, inside MULTI/EXEC its expected to provide exactly one reply - or else the client will block waiting for it.

It seems like the protocol contradicts itself and I actually suspect this is a bug inside Redis. I'm checking with redis-py in raw mode as the probably most common client backed by Redis itself, and it fails to parse the output correctly there as well.

Redis-py actually doesn't allow SUBSCRIBE inside MULTI. So I think all we can do is just accept that its behavior is not defined. 

---

As to the pipeline stuff, that's a valid issue. I fixed it already in my PR (but need to rollback and simplify some fixes I made for multi/exec that are actually not needed)
